core/state/trie_prefetcher.go

```diff
@@ -246,139 +309,307 @@ type subfetcher struct {
-		// Subfetcher already terminated, return a copy directly
-		if sf.trie == nil {
-			return nil
-		}
-		return sf.db.CopyTrie(sf.trie)
+// wait must only be called if [triePrefetcher] has not been closed. If this happens,
+// workers will not finish.
+func (sf *subfetcher) wait() {
+	if sf.to == nil {
+		// Unable to open trie
+		return
 	}
+	sf.to.wait()
 }
 
-// abort interrupts the subfetcher immediately. It is safe to call abort multiple
-// times but it is not thread safe.
 func (sf *subfetcher) abort() {
-	select {
-	case <-sf.stop:
-	default:
-		close(sf.stop)
+	if sf.to == nil {
+		// Unable to open trie
+		return
 	}
-	<-sf.term
+	sf.to.abort()
 }
 
-// loop waits for new tasks to be scheduled and keeps loading them until it runs
-// out of tasks or its underlying trie is retrieved for committing.
-func (sf *subfetcher) loop() {
-	// No matter how the loop stops, signal anyone waiting that it's terminated
-	defer close(sf.term)
+func (sf *subfetcher) skips() int {
+	if sf.to == nil {
+		// Unable to open trie
+		return 0
+	}
+	return int(sf.to.skips.Load())
+}
 
+func (sf *subfetcher) copies() int {
+	if sf.to == nil {
+		// Unable to open trie
+		return 0
+	}
+	return sf.to.copies
+}
+
+// trieOrchestrator is not thread-safe.
+type trieOrchestrator struct {
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+
+	sf *subfetcher
+
+	// base is an unmodified Trie we keep for
+	// creating copies for each worker goroutine.
+	//
+	// We care more about quick copies than good copies
+	// because most (if not all) of the nodes that will be populated
+	// in the copy will come from the underlying triedb cache. Ones
+	// that don't come from this cache probably had to be fetched
+	// from disk anyways.
+	base     Trie
+	baseLock sync.Mutex
+
+	outstandingRequests sync.WaitGroup
+	tasksAllowed        bool
+	skips               atomic.Int32 // Number of tasks skipped
+	pendingTasks        [][]byte
+	pendingTasksLock    sync.Mutex
+	wake                chan struct{}
+	loopTerm            chan struct{}
+
+	copies      int
+	copyChan    chan Trie
+	copySpawner chan struct{}
+}
+
+func newTrieOrchestrator(sf *subfetcher) *trieOrchestrator {
 	// Start by opening the trie and stop processing if it fails
+	var (
+		base Trie
+		err  error
+	)
 	if sf.owner == (common.Hash{}) {
-		trie, err := sf.db.OpenTrie(sf.root)
+		base, err = sf.db.OpenTrie(sf.root)
 		if err != nil {
 			log.Warn("Trie prefetcher failed opening trie", "root", sf.root, "err", err)
-			return
+			return nil
 		}
-		sf.trie = trie
 	} else {
-		trie, err := sf.db.OpenStorageTrie(sf.state, sf.owner, sf.root)
+		base, err = sf.db.OpenStorageTrie(sf.state, sf.owner, sf.root)
 		if err != nil {
 			log.Warn("Trie prefetcher failed opening trie", "root", sf.root, "err", err)
-			return
+			return nil
 		}
-		sf.trie = trie
 	}
 
-	// Trie opened successfully, keep prefetching items
+	// Create cancelable context
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
Could we get rid of cancellation here in favor of calling `Stop` on `BoundedWorkers` (to handle allowing all in-flight work to finish) and then terminate?

I think that should simplify a good bit of the code here.